Protocol Upgrade

samuelsadok opened the pull request:
Replace the user-oriented communication protocol with an improved machine oriented protocol.

The main differences are:
 - the device sends a JSON descriptor to announce it's parameters and functions to clients
 - CRC16 for UART
 - acknowledgements for parameter writes (and reads and function calls)

`test_communication.py` still provides the "p/v/c" commands but uses the new protocol internally. The tool now accepts a "--usb" and "--serial" parameter to select a specific device.

The old protocol is still operational on USB.

madcowswe committed:
C++ top level passthrough works

madcowswe committed:
comms hijack compiles

madcowswe committed:
ready to move state machine to cpp

madcowswe committed:
motor_thread pushed up to Axis

madcowswe committed:
formatting

samuelsadok committed:
move USB command handling to separate files, preliminary r, w and j commands

madcowswe committed:
add comments Oskar

samuelsadok committed:
fix typo

samuelsadok committed:
add python support for new protocol

samuelsadok committed:
Merge branch 'sam_protocol' of https://github.com/madcowswe/ODriveFirmware into sam_protocol

samuelsadok committed:
add back g and s commands, add read-only flag to JSON

madcowswe committed:
Merge branch 'uart' into cppify

 Conflicts:
	Makefile
	MotorControl/low_level.c
	MotorControl/low_level.h
	Src/freertos.c

madcowswe committed:
working after merge

madcowswe committed:
put printf float flags in LDFLAGS

madcowswe committed:
Push firmware into Firmware folder

madcowswe committed:
Push firmware into Firmware folder

madcowswe committed:
Merge branch 'uart' into cppify

madcowswe committed:
Merge branch 'uart' into cppify

samuelsadok committed:
Merge remote-tracking branch 'origin/cppify' into sam_protocol

samuelsadok committed:
fix messy merge

samuelsadok committed:
add binary protocol with CRC and acks

samuelsadok committed:
preliminary (untested) port of new protocol to python

madcowswe committed:
copy paste bug fix

samuelsadok committed:
[python] receive acks

samuelsadok committed:
Merge remote-tracking branch 'origin/devel' into sam_protocol

samuelsadok committed:
fix USB

madcowswe committed:
Merge branch 'uart' into cppify

samuelsadok committed:
almost make new binary protocol work with python (up to fetching and parsing the JSON)

samuelsadok committed:
Merge branch 'sam_protocol' of github.com:madcowswe/ODrive into sam_protocol

samuelsadok committed:
make float read/write work

madcowswe committed:
fix delay and timings

madcowswe committed:
add comments

madcowswe committed:
add comments

samuelsadok committed:
update test_communication.py to new backend, deferred USB processing, move CRC16 to stream based layer, some protocol refactoring

samuelsadok committed:
make remote precedure calls work

samuelsadok committed:
move Visual Studio Code settings to top level dir

samuelsadok committed:
python tool: add cmdline option to select a specific device (on USB or serial)

samuelsadok committed:
fix UART protocol, enable legacy protocol coexistence on USB port

samuelsadok committed:
update VS Code settings

samuelsadok committed:
minor comment clean-ups

samuelsadok committed:
filled complete endpoint table for new protocol
(ok we need an autogenerator for this)

samuelsadok committed:
change sync byte to 0xAA

kalvdans commented on Firmware/MotorControl/crc.hpp:
```diff
@@ -0,0 +1,66 @@
+template<typename T, unsigned POLYNOMIAL>
+static T calc_crc(T remainder, uint8_t value) {
+    constexpr T BIT_WIDTH = (CHAR_BIT * sizeof(T));
+    constexpr T TOPBIT = (1 << (BIT_WIDTH - 1));
```
Signed overflow if BIT_WIDTH is 32 or 64. Use `1ull << (BIT_WIDTH - 1)` or `(T)1 << (BIT_WIDTH - 1)` instead.

samuelsadok replied:
Good catch, fixed. For T=int32_t I don't see any issue though.

madcowswe replied:
I think in the int32_t case, the signed overflow is technically undefined (though in practice it's bit-wise equivalent to unsigned overflow, which is defined).

https://stackoverflow.com/a/16188846/1128910

madcowswe committed:
lift up c_cpp_properties.json

madcowswe committed:
Merge branch 'cppify' into sam_protocol

madcowswe committed:
some minor formating and naming

madcowswe committed:
Merge branch 'devel' into sam_protocol

samuelsadok committed:
make CRC work for 64-bit

madcowswe committed:
fix some issues with moving anticogging

madcowswe committed:
add comments

madcowswe committed:
add comments

madcowswe committed:
add comments

madcowswe committed:
add explore_odrive.py

samuelsadok committed:
Merge branch 'devel' of github.com:madcowswe/ODrive into sam_protocol

samuelsadok committed:
fix osSemaphoteWait return type

samuelsadok committed:
miscellaneous protocol updates

add back support for legacy protocol on UART
add defines in commands.h for easy selection of protocols
new protocol: use semaphores for sending and drop response after 10ms
minor typing fixes

samuelsadok committed:
miscelleneous python updates

accept arguments in explore_odrive.py (like in test_communication.py)
implement --discover argument
fix spurious property fetch
fix read-only properties
better serial port detection for windows
revert demo.py

samuelsadok committed:
update communication documentation

samuelsadok committed:
fix doc formatting

samuelsadok committed:
Merge branch 'devel' of github.com:madcowswe/ODrive into sam_protocol

samuelsadok committed:
fix default GPIO mode

madcowswe committed:
split .vscode settings between firmware and tools folder

madcowswe committed:
add VSCode workspace file

madcowswe committed:
autoformat low_level.c

madcowswe committed:
Merge branch 'devel' into sam_protocol

madcowswe committed:
fix some merge errors

madcowswe committed:
remove old test_communication.py

madcowswe committed:
add c++ include paths to VSCode

madcowswe committed:
Improve documentation, allow UART on v3.3 and higher only

madcowswe committed:
add IPython support for explore_odrive.py

madcowswe committed:
update changelog

madcowswe committed:
re-remove file inadvertently reintroduced

madcowswe committed:
minor change
